pkg/topology/builder.go

```diff
@@ -0,0 +1,652 @@
+func (b *Builder) populateTrafficSplitsAuthorizedIncomingTraffic(topology *Topology) {
+	loopDetected := make(map[*Service][]*TrafficSplit)
+
+	for _, svc := range topology.Services {
+		for _, ts := range svc.TrafficSplits {
+			pods, err := b.getIncomingPodsForTrafficSplit(ts, map[Key]struct{}{})
+			if err != nil {
+				loopDetected[svc] = append(loopDetected[svc], ts)
+				b.Logger.Errorf("Unable to get incoming pods for TrafficSplit %s/%s: %v", ts.Namespace, ts.Name, err)
+
+				continue
+			}
+
+			ts.Incoming = pods
+		}
+	}
+
+	// Remove the TrafficSplits that causes a loop.
+	for svc, tss := range loopDetected {
+		for _, loopTs := range tss {
+			for i, ts := range svc.TrafficSplits {
+				if ts == loopTs {
+					svc.TrafficSplits = append(svc.TrafficSplits[:i], svc.TrafficSplits[i+1:]...)
+					break
+				}
+			}
+		}
+	}
+}
+
+func (b *Builder) getIncomingPodsForTrafficSplit(ts *TrafficSplit, visited map[Key]struct{}) ([]*Pod, error) {
+	keyTS := Key{ts.Name, ts.Namespace}
+	if _, found := visited[keyTS]; found {
+		return nil, fmt.Errorf("circular reference detected on traffic split %s/%s in service %s/%s", ts.Namespace, ts.Name, ts.Service.Namespace, ts.Service.Name)
+	}
+
+	visited[keyTS] = struct{}{}
+
+	var union []*Pod
+
+	for _, backend := range ts.Backends {
+		backendPods, err := b.getIncomingPodsForService(backend.Service, mapCopy(visited))
+		if err != nil {
+			return nil, err
+		}
+
+		union = unionPod(backendPods, union)
+
+		if len(union) == 0 {
+			return union, nil
+		}
+	}
+
+	return union, nil
+}
+
+func (b *Builder) getIncomingPodsForService(svc *Service, visited map[Key]struct{}) ([]*Pod, error) {
+	var union []*Pod
+
+	if len(svc.TrafficSplits) == 0 {
+		var pods []*Pod
+
+		for _, tt := range svc.TrafficTargets {
+			for _, source := range tt.Sources {
+				pods = append(pods, source.Pods...)
+			}
+		}
+
+		return pods, nil
+	}
+
+	for _, ts := range svc.TrafficSplits {
+		tsPods, err := b.getIncomingPodsForTrafficSplit(ts, visited)
+		if err != nil {
+			return nil, err
+		}
+
+		union = unionPod(tsPods, union)
+
+		if len(union) == 0 {
+			return union, nil
+		}
+	}
+
+	return union, nil
+}
+
+func unionPod(pods1, pods2 []*Pod) []*Pod {
+	var union []*Pod
+
+	if pods2 == nil {
+		return pods1
+	}
+
+	p := map[*Pod]bool{}
+
+	for _, pod := range pods1 {
+		p[pod] = true
+	}
+
+	for _, pod := range pods2 {
+		if _, found := p[pod]; found {
+			union = append(union, pod)
+		}
+	}
+
+	return union
+}
+
+// buildTrafficTargetSources retrieves the Pod IPs for each Pod mentioned in a source of the given TrafficTarget.
+// If a Pod IP is not yet available, the pod will be skipped.
+func (b *Builder) buildTrafficTargetSources(res *resources, t *Topology, tt *access.TrafficTarget) []ServiceTrafficTargetSource {
+	sources := make([]ServiceTrafficTargetSource, len(tt.Sources))
+
+	for i, source := range tt.Sources {
+		srcSaKey := Key{source.Name, source.Namespace}
+
+		pods := res.PodsBySa[srcSaKey]
```

Comment:
```suggestion
		pods := res.PodsByServiceAccounts[srcSaKey]
```

I think this will help :)